draftjs_exporter/engines/string.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import absolute_import, unicode_literals
+
+from draftjs_exporter.engines.base import DOMEngine
+
+try:
+    # Python 3.2 and above.
+    from html import escape
+except ImportError:
+    import cgi
+
+    def escape(s):
+        # Force quote escaping in Python 2.7.
+        return cgi.escape(s, quote=True).replace('\'', '&#x27;')
+
+# http://w3c.github.io/html/single-page.html#void-elements
+# https://github.com/html5lib/html5lib-python/blob/0cae52b2073e3f2220db93a7650901f2200f2a13/html5lib/constants.py#L560
+VOID_ELEMENTS = set([
+    'area',
+    'base',
+    'br',
+    'col',
+    'embed',
+    'hr',
+    'img',
+    'input',
+    'link',
+    'meta',
+    'param',
+    'source',
+    'track',
+    'wbr',
+])
+
+
+class DOMString(DOMEngine):
+    """
+    string concat implementation of the DOM API.
+    """
+
+    @staticmethod
+    def create_tag(type_, attr=None):
+        return {
+            'type': type_,
+            'attr': attr,
+            'children': [],
+        }
+
+    @staticmethod
+    def append_child(elt, child):
+        # This check is necessary because the current wrapper_state implementation
+        # has an issue where it inserts elements multiple times.
+        # This must be skipped for text, which can be duplicated.
+        is_existing_ref = isinstance(child, dict) and child in elt['children']
+        if not is_existing_ref:
+            elt['children'].append(child)
+
+    @staticmethod
+    def render_attrs(attr):
+        attrs = [' %s="%s"' % (a, escape(attr[a])) for a in attr]
+        return ''.join(sorted(attrs))
```

Comment:
This should be faster, especially because we get rid of the useless sorting. In Python ≤ 3.5, this may generate different HTML from a run to another one, so it may affect the tests if you’re not using `assertHTML`.
```python
@staticmethod
def render_attrs(attr):
    return ''.join([' %s="%s"' % (k, escape(v)) for k, v in attr.items()])
```

Reply:
I think it's quite valuable to have the same order of attributes across the board, so I'll keep the sorting for now. BeautifulSoup / html5lib and lxml also output attributes sorted. 

I'm happy for this to be revisited later, but yep there'll also need to be some rework in the tests because there is no such thing as `assertHTMLEqual` in vanilla Python AFAIK.